doc/preprocessor.rst

```diff
@@ -0,0 +1,130 @@
+--------
+Instead, you can provide a callable in the argument ``preprocessor``.
+Then the estimator will accept indicators of points instead of points.
+Under the hood, the estimator will call this callable on the indicators you
+provide as input when fitting, predicting etc...
+Using a callable can be really useful to represent lazily a dataset of
+images stored on the file system for instance.
+The callable should take as an input an array-like, and return a 2D
+array-like. For supervised learners it will be applied on the whole array of
+indicators at once, and for weakly supervised learners it will be applied
+on each column of the array of tuples.
+
+Example with a supervised metric learner:
+
+The callable should take as input an array-like, and return a 2D array-like.
+
+>>> def find_images(arr):
+>>>     X = np.array([[-0.7 , -0.23],
+>>>                   [-0.43, -0.49],
+>>>                   [ 0.14, -0.37]])  # array of 3 samples of 2 features
+>>>     result = []
+>>>     for img_path in arr:
+>>>         result.append(X[int(img_path[3:5])])
+>>>         # transforms 'img01.png' into X[1]
+>>>     return np.array(result)
+>>> images_paths = ['img01.png', 'img00.png', 'img02.png']
+>>> y = np.array([1, 0, 1])
+>>>
+>>> nca = NCA(preprocessor=find_images)
+>>> nca.fit(images_paths, y)
+>>> # under the hood preprocessor(indicators) will be called
+
+
+Example with a weakly supervised metric learner:
+
+The given callable should take as input an array-like, and return a
+2D array-like. It will be called on each column of the input tuples of
+indicators.
+
+>>> def find_images(arr):
+>>>     X = np.array([[-0.7 , -0.23],
+>>>                   [-0.43, -0.49],
+>>>                   [ 0.14, -0.37]])  # array of 3 samples of 2 features
+>>>     result = []
+>>>     for img_path in arr:
+>>>         result.append(X[int(img_path[3:5])])
+>>>         # transforms 'img01.png' into X[1]
+>>>     return np.array(result)
+>>> pairs_images_paths = [['img02.png', 'img00.png'],
+>>>                       ['img01.png', 'img00.png']]
+>>> y_pairs = np.array([1, -1])
+>>>
+>>> mmc = NCA(preprocessor=find_images)
+>>> mmc.fit(pairs_images_paths, y_pairs)
+>>> # under the hood preprocessor(pairs_indicators[i]) will be called for each
+>>> # i in [0, 1]
+
+
+.. note:: Note that when you fill the ``preprocessor`` option, it allows you
+ to give more compact inputs, but the classical way of providing inputs
+ stays valid (2D array-like for ``X`` for supervised learners and 3D
+ array-like of tuples for weakly supervised learners).
+
+ Example: This would work:
+
+ >>> from metric_learn import MMC
+ >>> X = np.array([[-0.7 , -0.23],
+ >>>               [-0.43, -0.49],
+ >>>               [ 0.14, -0.37]])  # array of 3 samples of 2 features
+ >>> pairs = np.array([[[ 0.14, -0.37], [-0.7 , -0.23]],
+ >>>                   [[-0.43, -0.49], [-0.7 , -0.23]]])
+ >>> y_pairs = np.array([1, -1])
+ >>>
+ >>> mmc = MMC(preprocessor=X)
+ >>> mmc.fit(pairs, y_pairs)
```

Comment:
Maybe be explicit here that the `X` array is not used at all in this case.

Reply:
I agree, I replaced it with a "not implemented yet" preprocessor, and wrote that the code would work anyways